### ISSUE TITLE
Behaviour of py-*: how to pass event

### ISSUE THREAD

**Issue by dj-fiorex** [pyscript/pyscript#835]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

Hello, i already used the py-* functionality to call a python function like this:

```
<input
       class="form-control"
       id="myfile"
       type="file"
       py-change="form_field_changed"
      />
...


 <py-script >
def form_field_changed(event):
</py-script>
```

Now this doesn't work anymore.
I noted that if i put brackets on function call like this form_field_changed() then the function is executed but comply that it require an argument (event)

is the Behaviour changed?
how can i access the event object?

Thanks

### What browsers are you seeing the problem on? (if applicable)

Firefox, Chrome, Safari, Microsoft Edge, Other

**Comment by JeffersGlass** [comment]
This behavior was changed as part of #686, which was included in the 2022.09.1 release of PyScript which was published last week. If you're linking to `https://pyscript.net/latest/pyscript.js`, then that's what probably broke your code.

The new syntax is:

```html
<input
  py-change="form_field_changed()"
  ...
/>

<py-script>
    def form_field_changed():
        ...
</py-script>
```

This change was made to more closely mirror [JavaScript's onevent syntax](https://developer.mozilla.org/en-US/docs/Web/Events/Event_handlers).

Note that PyScript's syntax does not currently make the `event` object available to the event handling method - you'll want to include logic for grabbing the data from the input in your `form_field_changed` function.

**Comment by dj-fiorex** [comment]
ok thank you for the explanation!
Maybe the best thing to do is to have some examples that highlight the usage of py-* attributes. Where can i create it?

**Comment by tedpatrick** [comment]
Yes, we need a way to get the event passed cleanly into the function. Looks like an area of work.

**Comment by dj-fiorex** [comment]
I'm ready to give some help in my free time, where can i look to get an overview of the event handler?

**Comment by JeffersGlass** [comment]
Thanks @dj-fiorex! That would be swell. The logic for the `py-*event` handling is in [pyscript.ts](https://github.com/pyscript/pyscript/blob/66b57bf812dcc472ed6ffee075ace5ced89bbc7c/pyscriptjs/src/components/pyscript.ts#L119-L260). 

Perhaps writing up an explanation or reference-guide for the [docs section](https://github.com/pyscript/pyscript/tree/main/docs) would be the best place for that info to go, if you're game to write it up. @marimeireles may have thoughts as to the best home for that info.

For what it's worth, not every possible event is currently captured (though all the 'common' ones are) - #801 has some details. 

---

For what it's worth, and slightly orthogonal to your question here, events _are_ available if you add the event listener separately:

```html
<!-- py-*event syntax doesn't current pass event nor "this"-->
<button id="one" py-click="say_hello()">No Event with Py-click</button>
<py-script>
  from js import console,

  def say_hello():
    console.log("HELLO!")
</py-script>

<!-- but pyodide.ffi.wrappers.add_event_listener or Element.addEventListener do-->
<button id="two">add_event_listener passes event</button>
<py-script>
  from js import console, document
  from pyodide.ffi.wrappers import add_event_listener

  def hello_args(*args):
    console.log(f"Hi! I got some args! {args}")

  add_event_listener(document.getElementById("two"), "click", say_goodbye)
</py-script>
```

**Comment by tedpatrick** [comment]
I am working on this at 2 levels:

1. How do we subscribe py-{event} at startup
2. How do we support adding a py-{event} attribute at runtime after PyScript has initialized (mutation observer)

```html
Case 1: <button py-click="foo()">ClickMe</button>             SUPPORTED
Case 2: <button py-click="foo">ClickMe</button>               PROPOSED
```

In case 2, if the target function has 1 argument, it gets passed the event. If no arguments, no event.

**Comment by tedpatrick** [comment]
You can also use a create_proxy
```html
<html>
<head>
    <link rel="stylesheet" href="https://pyscript.net/unstable/pyscript.css" />
    <script defer src="https://pyscript.net/unstable/pyscript.js"></script>
</head>
<body>
    <button id="two">add_event_listener passes event</button>
<py-script>
  from pyodide.ffi import create_proxy

  def hello_args(*args):
      console.log(f"Hi! I got some args! {args}")

  Element('two').element.addEventListener("click", create_proxy(hello_args))
</py-script>
</body>
</html>
```

**Comment by JeffersGlass** [comment]
+1 to sometime having a solution that works when attributes are added after initialization - from the amount of times this has come up recently, I suspect PyScript will ultimately be one large mutation observer. And not in a bad way!

> Case 1: <button py-click="foo()">ClickMe</button>             SUPPORTED
> Case 2: <button py-click="foo">ClickMe</button>               PROPOSED

-1 to case 2: that was the sole behavior in 2022.06.1, and we specifically moved away from it with PR #686, because it's not how script onevent syntax works. If it's possible:

```
Case 3: <button py-click="foo(event)">ClickMe</button>       POSSIBLE??
```
would be closest, I think, though I have no idea how to implement it.

**Comment by tedpatrick** [comment]
That syntax is what several leading JS frameworks do today, it isn't standard but it is the common syntax for assigning attributes function values by name. [vue](https://vuejs.org/guide/essentials/event-handling.html#method-handlers) [react](https://reactjs.org/docs/handling-events.html) [svelte](https://svelte.dev/tutorial/component-events)

Much is possible, the key is what interface is best for building apps from a developer's perspective. I personally would use this all the time. And what if it worked depending on the target method's arguments? I believe there is merit in supporting both.

```html
<button py-click="foo">ClickMe</button>
```

```html
<button py-click="foo()">ClickMe</button>
```

**Comment by marimeireles** [comment]
Hey @dj-fiorex, my input for docs is that we may start introducing real content to the [Reference](https://docs.pyscript.net/latest/reference/index.html) part.
I guess it'd be nice to have a) an explanation on how to use the function, b) a list of the possible events, c) maybe @tedpatrick's example on how to use it through proxy is useful to have it there?
Thank you for offering to contribute 🎉 that's awesome :)

**Comment by tedpatrick** [comment]
And now for something completely different...

In exploring how to address the issues in py-event attributes, I took a detour and ended up somewhere interesting.

There are 4 core issues:
1. We need to get the `event` object into python for event handlers
2. It isn't clear how to expose a python function in JS
3. We need to support adding/removing these handlers at runtime (py-event is wired up once at startup)
4. We need an easy way to subscribe to events in python.

API (does not work without framework modifications)
```html
<body style="padding:2em;">
    <button onclick="py.foo()">call foo on click</button><br><br>
    <button onclick="py.boo(event)">call boo on click w event</button><br><br>
    <button onclick="py.loo(event, 123)">call loo on click w event + 123</button><br><br>
    <button onclick="py.nothing(event)">call nothing on click w event</button><br><br>
    <button id="empty_button">empty_button with mousemove event</button><br><br>
    <py-script>
import js

@handler
def foo():
    js.console.log("foo called")

@handler
def boo(e):
    js.console.log("boo called with event")
    js.console.log( e )

@handler
def loo(*e):
    js.console.log("loo called with *event")
    js.console.log( e[0] )
    js.console.log( e[1] )

@handler(event="click")
def doc_click(*e):
    js.console.log("doc_click called with *event")
    js.console.log( e[0] )

@handler( event="mousemove", element=Element("empty_button").element )
def button_mousemove(*e):
    js.console.log("button_mousemove called with *event")
    js.console.log( e[0] )

    </py-script>
</body>
```

This is using the on[event] attribute syntax in JS but it adds in a global `py` object to allow writing event handlers like so :
 - `onclick="py.foo()"` 
 - `onclick="py.loo(event)"` 
 - `onclick="py.goo(42,event,1,2,3)"` 

within all on[event] attributes, the `event` object exists within the call. So we can use `event` and write it directly into a call to a python handler. Since this argument interface is shared across js and python, it is easy to wire this in seamlessly.

As for exposing python methods into the `py` object, there is a `@handler` decorator that can simply add a function or wire up the event handler without using html/js at all.

Here we expose the button_mousemove handler to js and subscribe it to the mousemove event for a button element.
```python
@handler( event="mousemove", element=Element("empty_button").element )
def button_mousemove(*e):
    js.console.log("button_mousemove called with *event")
```

Here is a [working version for review](https://shimmering-naiad-03376f.netlify.app/). Open the JS console to see this working fully. [Source is here](https://github.com/tedpatrick/pyscript_events_poc/blob/main/index.html).

The irony is in exploring py-event, there might be a cleaner, simpler way to manage events within a PyScript application.

Note, this is VERY EARLY and VERY POC. If we were to move forward along these lines, the naming and defaults would change but I think it feels right.

Feedback welcome.

**Comment by pmp-p** [comment]
i ~~like~~ love the py.* proxy for event ( i already use that in pygame-script), but i don't think `event` has its place in arguments it should be where it is expected to be ie : `js.window.event`

while inline onclick arguments should be fully and strictly in javascript space. That would allow to pass `this` properly instead of setting builtins.this  before reaching python side.

for solely accessing event the method `Element('two').element.addEventListener("click", create_proxy(hello_args))` described earlier is the way.

**Comment by antocuni** [comment]
> i ~like~ love the py.* proxy for event ( i already use that in pygame-script), but i don't think `event` has its place in arguments it should be where it is expected to be ie : `js.window.event`

According to [Mozilla's docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/event) `window.event` is deprecated, isn't it?

**Comment by pmp-p** [comment]
> According to [Mozilla's docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/event) window.event is deprecated, isn't it?

Sure but the window.event on python side is empty, so it's safe to use it for what it is used for : passing event for inline onclick=click() 

doing onclick=click(event) would imply addressing window.event ... So if it vanishes how one would get the event ?

At least window.event is a single place in the proxy  to maintain and to adapt to future ways.
it's also easy to make https://pygame-web.github.io/showroom/pyscript/unify.html

**Comment by antocuni** [comment]
> Sure but the window.event on python side is empty, so it's safe to use it for what it was/is used for : passing event for argument-less inline onclick=click()

I don't understand. Are you proposing that we should write to `window.event`?
If so, I don't think it's not safe at all. Even on the Python side, `js.window` is a reference to the real `window` object on the JS side, so if we try to write a `.event` attribute on it, I'm sure things will explode.

> doing onclick=click(event) would imply addressing window.event ... So if it vanishes suddenly after deprecation how one would get the event in old code ?

I admit that I am having troubles to understand this sentence of the mozilla docs:
> you should instead use the [Event](https://developer.mozilla.org/en-US/docs/Web/API/Event) passed into the event handler function.

does it mean that you should not use the `onclick=...` HTML attribute but that you have to manually use `addEventListener` if you want to access the `event`?

**Comment by tedpatrick** [comment]
Window.event is deprecated but it has very wide support
<https://developer.mozilla.org/en-US/docs/Web/API/Window/event#browser_compatibility>
and its removal would break the web.

The @handler( event="click" ) decorator uses addEventListener internally so
this essentially aligns with how events work in the DOM Event spec. You can
use on[event] handlers and you get the existing behavior or you can use
addEventListener directly or via the @handler decorator.

on[event] attributes have limitations in that one 1 can exist for a single
event type whereas addEventListener events can have many listeners and
support event options.

In py-event attributes, there is currently no way to obtain the event and
pass it to python. Also, py-event attributes are only set at startup and
lack any runtime flexibility.

The interesting part here is that these are actually independent of
on[event] does not conflict with py-event or vice versa.

**Comment by pmp-p** [comment]
> If so, I don't think it's not safe at all. Even on the Python side, js.window is a reference to the real window object on the JS side, so if we try to write a .event attribute on it, I'm sure things will explode. 

indeed it can have side effects and it's wasted cycle to send it back to js,  why not just set "builtins.event" then only on python side  ( i updated https://pygame-web.github.io/showroom/pyscript/unify.html accordingly ) ?

ps @antocuni: it's "black -x" formatted now,  and maybe that's why i wanted json in pyconfig instead of toml ;)

**Comment by antocuni** [comment]
> indeed it can have side effects and it's wasted cycle to send it back to js, why not just set "builtins.event" then only on python side 

well, relying on global state seems suboptimal and a certain source of bug. I would strongly prefer a solution in which we pass `event` as a function argument (even if on the JS side is a global -- it's not a good reason for continuing doing wrong things).

Example of things which can go wrong with a global/window/builtins `event`:

```python
def on_click():
    assert event.type == "click" # guaranteed to work
    await fetch_something()
    assert event.type == "click" # might fail if another event handler was fired in the meantime
```

**Comment by dj-fiorex** [comment]
Hello guys, i'm a web developer for 90% of my time and 10% python, and i can say that in the web space the event handler receive an event as an argument. 
What @tedpatrick suggests is the way to go in my opinion

**Comment by pauleveritt** [comment]
Small note...I agree with @antocuni about the event argument being deprecated. My IDE is yelling at me when it is used. 😀 Yes it is still implemented. But if it's a bad practice, it's a bad practice.

**Comment by antocuni** [comment]
> Small note...I agree with @antocuni about the event argument being deprecated. My IDE is yelling at me when it is used. grinning Yes it is still implemented. But if it's a bad practice, it's a bad practice.

I suppose you mean the `event `global variable such as in `<button onclick="console.log(event)">`?

**Comment by marimeireles** [comment]
I guess this will be fixed once https://github.com/pyscript/pyscript/pull/1200 is merged

**Comment by WebReflection** [comment]
I know in classic this discussion had reasons to exist but with current PyScript, thisis the contract:

```
Case 1: <button py-click="foo()">ClickMe</button>             NOT SUPPORTED
Case 2: <button py-click="foo">ClickMe</button>               WORKS
```

For the `@when` use case and explicit trigger we should have a different and well defined issue, imho, so I am closing this as it won't get fixed in classic and it's pointless to argue about this in current PyScript as we discussed this long time and the current state is not going to change.
